### PR TITLE
ci: add per-module integration coverage gate

### LINKED ISSUE #31: test: add unit tests for check_module_coverage_floor.py
## Problem

`scripts/check_module_coverage_floor.py` is untested production CI logic. The script's correctness gates every integration run, but none of its core logic has test coverage.

## Risk

Silent mis-parses are easy to introduce in:

- `parse_report` — ANSI-stripping, the GH-Actions tab-prefix heuristic, and the ROW_RE regex all have edge cases
- `evaluate` — tolerance boundary conditions (e.g. `actual + tolerance == floor` should pass, `actual + tolerance < floor` should fail)
- `missing_from_report` — the `(repo_root / module).exists()` branch exercises filesystem state that is easy to mis-test

## Acceptance Criteria

- [ ] `tests/scripts/test_check_module_coverage_floor.py` created
- [ ] `parse_report`: tests for clean ASCII report, ANSI-escaped report, tab-prefixed GH Actions lines, lines that don't match ROW_RE (e.g. TOTAL row, blank lines), and a module whose coverage is an integer (no decimal)
- [ ] `evaluate`: regression detected when `actual < floor - tolerance`, passes when `actual == floor - tolerance`, passes when `actual > floor`, new-module below `new_module_min` detected, new-module at exactly `new_module_min - tolerance` passes, baseline module absent from report but file still exists flagged, baseline module absent from report but file deleted → not flagged
- [ ] `main` (integration): baseline not found → exit code 2, report not found → exit code 2, all clean → exit code 0, regressions present → exit code 1
- [ ] All assertions use concrete expected values, not `isinstance` or `>= 0` (per T1/T9)
- [ ] Tests use `tmp_path` fixtures, not hardcoded paths

### LINKED ISSUE #32: fix: handle malformed baseline JSON with clean error in check_module_coverage_floor.py
## Problem

`load_baseline` calls `json.loads` without catching `json.JSONDecodeError`. A malformed `integration_module_floor_baseline.json` produces an unhandled traceback instead of a structured error message, making CI output harder to diagnose.

```python
def load_baseline(path: Path) -> dict[str, Any]:
    data = json.loads(path.read_text(encoding="utf-8"))  # ← bare, no error handling
```

This is inconsistent with `validate_inputs`, which already follows the clean `"ERROR: ... → return 2"` pattern.

## Fix

Wrap `json.loads` and `path.read_text` in `load_baseline` (or promote the call into `main`) so that:

1. `json.JSONDecodeError` → `print(f"ERROR: baseline file is not valid JSON: {path}\n{e}")` → return `2`
2. The function signature changes to return `dict[str, Any] | None` **or** the error is raised as a typed exception caught in `main`

## Acceptance Criteria

- [ ] `load_baseline` never lets `json.JSONDecodeError` propagate uncaught to the user
- [ ] Error output matches the `"ERROR: ..."` format used elsewhere in the script
- [ ] Exit code is `2` (input error) not `1` (gate failure)
- [ ] Test added: `tmp_path` baseline with `{not json}` content → `main()` returns `2` and prints `ERROR:`

### LINKED ISSUE #33: fix: anchor repo_root to script location instead of cwd in check_module_coverage_floor.py
## Problem

`main()` passes `repo_root=Path.cwd()` to `evaluate()`:

```python
regressions, low_new_modules, missing_from_report = evaluate(
    ...
    repo_root=Path.cwd(),   # ← breaks if invoked from any directory other than repo root
)
```

`evaluate` uses this to check `(repo_root / module).exists()` when a baseline module disappears from the coverage report. If the script is invoked from a subdirectory (e.g. `cd scripts && python check_module_coverage_floor.py ...`), the path check silently fails: existing modules are not found, so they are excluded from `missing_from_report`, hiding real regressions.

## Fix

Anchor to the script's own location, which is always known:

```python
# In evaluate() call inside main():
repo_root=Path(__file__).resolve().parents[1],  # scripts/ → repo root
```

Or expose it as an optional CLI argument (`--repo-root`, default to the above) for flexibility in unusual CI layouts.

## Acceptance Criteria

- [ ] `repo_root` is no longer derived from `cwd`
- [ ] Default is `Path(__file__).resolve().parents[1]` or an equivalent anchor
- [ ] Optional `--repo-root` CLI flag added (useful for testing and non-standard layouts)
- [ ] Test: invoke `main()` with a working directory different from the repo root and verify `missing_from_report` still correctly identifies a present-but-unreported module

### LINKED ISSUE #34: fix: print_report displays nominal min instead of effective floor for new-module failures
## Problem

`print_report` shows the raw `new_module_min` value:

```python
print(f"  - {module}: actual={actual:.1f}% min={new_module_min:.1f}%")
```

But the gate actually enforces `new_module_min - tolerance` (71.4% with current defaults of 71.9 / 0.5). A developer whose module is at **71.5%** will see:

```
  - src/file_organizer/foo.py: actual=71.5%  min=71.9%
```

and be confused — they beat 71.4% (the actual threshold) but the output implies they are 0.4pp short. This creates noise and erodes trust in the gate.

The same inconsistency appears in the regression path, where `floor` is displayed but the effective floor is `floor - tolerance`.

## Fix

Display the effective threshold in failure output:

```python
# For new modules:
effective_min = new_module_min - tolerance
print(f"  - {module}: actual={actual:.1f}%  effective_min={effective_min:.1f}%  (nominal={new_module_min:.1f}%  tolerance={tolerance:.1f}%)")

# For regressions:
effective_floor = floor - tolerance
print(f"  - {module}: actual={actual:.1f}%  effective_floor={effective_floor:.1f}%  (baseline={floor:.1f}%  tolerance={tolerance:.1f}%)")
```

Or at minimum add `(±{tolerance:.1f}% tolerance)` to the existing line.

## Acceptance Criteria

- [ ] Failure output shows the value a module actually needs to reach to pass
- [ ] The nominal baseline/min and tolerance are still visible for context
- [ ] A module at exactly `floor - tolerance` shows as passing (existing behaviour preserved)
- [ ] Test: assert output string for a regression case contains the effective floor, not the raw baseline

### LINKED ISSUE #35: fix: missing_from_report error message should guide the developer to resolution
## Problem

When a baseline module is missing from the coverage report, the script emits:

```
Baseline modules missing from coverage report:
  - src/file_organizer/foo.py
FAIL: per-module integration coverage gate failed
```

This tells the developer *what* happened but not *why* or *what to do*. The two root causes have different fixes:

1. **Module lost all integration test coverage** → add tests
2. **Module was renamed or moved** → update the baseline

Without guidance, developers are likely to delete the baseline entry (silently dropping the floor) rather than restore coverage.

## Fix

Augment the `missing_from_report` block in `print_report` with actionable guidance:

```
Baseline modules missing from coverage report:
  - src/file_organizer/foo.py

These modules exist on disk but were not reported by pytest-cov.
This usually means they have lost all integration test coverage.
To fix:
  1. Add integration tests so the module appears in the report, OR
  2. If the module was renamed/moved, update the baseline:
       scripts/coverage/integration_module_floor_baseline.json
```

## Acceptance Criteria

- [ ] `print_report` emits the guidance block whenever `missing_from_report` is non-empty
- [ ] Guidance distinguishes "lost coverage" from "renamed/moved"
- [ ] Baseline file path referenced in the message is derived from the actual `--baseline-path` argument, not hardcoded
- [ ] Test: assert the guidance text appears in captured stdout when `missing_from_report` is non-empty

### LINKED ISSUE #37: fix: per-module coverage check silenced when global gate or tests fail in CI
## Problem

The CI step is:

```yaml
run: |
  set -o pipefail
  pytest ... --cov-fail-under=71.9 ... | tee /tmp/integration-coverage-report.txt
  python scripts/check_module_coverage_floor.py ...
```

GitHub Actions runs steps under `bash -e` (exit-on-error). With `set -o pipefail`, a non-zero exit from pytest (whether from test failures **or** from `--cov-fail-under=71.9`) causes the pipe to exit non-zero, which under `-e` aborts the step. `check_module_coverage_floor.py` never runs.

This is worst-case behaviour: per-module diagnostics are most valuable precisely when something regressed, but the check is skipped in exactly that scenario.

## Fix

Decouple the two checks so the per-module script always runs when a coverage report was produced:

```yaml
- name: Run integration tests
  id: integration
  continue-on-error: true
  run: |
    set -o pipefail
    pytest tests/ -m "integration" \
      --strict-markers \
      --cov=file_organizer --cov-branch \
      --cov-report=term-missing \
      --cov-report=xml \
      --timeout=60 \
      --override-ini="addopts=" \
      | tee "$RUNNER_TEMP/integration-coverage-report.txt"

- name: Check global integration coverage floor
  run: |
    # Re-check the global threshold from the saved XML report
    coverage report --fail-under=71.9

- name: Check per-module integration coverage floors
  run: |
    python scripts/check_module_coverage_floor.py \
      --report-path "$RUNNER_TEMP/integration-coverage-report.txt" \
      --baseline-path scripts/coverage/integration_module_floor_baseline.json \
      --new-module-min 71.9 \
      --tolerance 0.5
```

Or alternatively, use `|| true` on the pytest pipe and re-assert the global gate as a separate `coverage report` invocation so both checks always execute and produce output.

## Acceptance Criteria

- [ ] Per-module check runs even when pytest exits non-zero due to coverage failure
- [ ] Global 71.9% gate is still enforced (not dropped)
- [ ] If pytest exits non-zero due to **test failures** (not coverage), the per-module check is skipped (no report to parse)
- [ ] CI step failure messages clearly distinguish "tests failed" vs "global coverage floor missed" vs "per-module floor missed"
- [ ] Both checks must individually gate the job (step fails → job fails)

### LINKED ISSUE #38: fix: use RUNNER_TEMP instead of /tmp for integration coverage report artifact in CI
## Problem

The CI step writes the coverage report to a hardcoded path:

```yaml
pytest ... | tee /tmp/integration-coverage-report.txt
```

`/tmp` is shared across all processes on a runner. If GitHub ever schedules two jobs on the same host concurrently (matrix runs, re-runs, or self-hosted runners with high parallelism), the second job overwrites the first job's report mid-read, corrupting the per-module check.

`$RUNNER_TEMP` is GitHub Actions' per-job scratch directory, guaranteed unique per job execution.

## Fix

Replace the hardcoded path in both the `tee` command and the `--report-path` argument:

```yaml
run: |
  set -o pipefail
  pytest tests/ -m "integration" \
    --strict-markers \
    --cov=file_organizer --cov-branch --cov-fail-under=71.9 \
    --cov-report=term-missing --cov-report=xml \
    --timeout=60 \
    --override-ini="addopts=" \
    | tee "$RUNNER_TEMP/integration-coverage-report.txt"
  python scripts/check_module_coverage_floor.py \
    --report-path "$RUNNER_TEMP/integration-coverage-report.txt" \
    --baseline-path scripts/coverage/integration_module_floor_baseline.json \
    --new-module-min 71.9 \
    --tolerance 0.5
```

## Acceptance Criteria

- [ ] `/tmp/integration-coverage-report.txt` replaced with `$RUNNER_TEMP/integration-coverage-report.txt` in `ci.yml`
- [ ] Both the `tee` destination and `--report-path` use the same variable
- [ ] No other `/tmp/` references remain in the workflow file for artifacts that could collide across concurrent jobs

### PR DESCRIPTION
**Pull request by curdriceaurora**
## Why
The global integration coverage gate catches aggregate regressions, but it can miss localized drops in specific modules. This PR preserves the repo-level bar and adds a robust per-module non-regression gate, with clearer diagnostics and local/CI parity.

## What Changed
- Added/expanded `scripts/check_module_coverage_floor.py` to enforce per-module floors from pytest-cov term output.
- Added strong error handling for malformed/invalid baseline input with controlled `ERROR:` output and exit code `2`.
- Anchored module existence checks to repo root (script-relative) instead of `cwd`, with optional `--repo-root` override.
- Improved failure output to show effective thresholds (`nominal - tolerance`) for both regressions and new-module checks.
- Added actionable guidance for `missing_from_report` failures, including baseline path in output.
- Updated policy resolution so baseline `policy` values are used by default (`new_module_min_percent`, `tolerance_percent`) with CLI override support.
- Hardened baseline module-floor parsing failures into controlled CLI errors.

## CI / Local Parity
- Updated integration workflow to run as explicit phases:
  1. run integration tests and capture coverage report
  2. run per-module gate
  3. run global coverage gate
  4. enforce outcomes with explicit final failure reasons
- Switched integration report temp artifact to `$RUNNER_TEMP` (instead of hardcoded `/tmp`).
- Updated `scripts/run-local-ci.sh` `run_integration()` to mirror CI behavior (per-module gate + global gate).

## Tests
- Expanded `tests/ci/test_module_coverage_floor.py` to cover:
  - parser behavior (ANSI, GH tab-prefix, integer coverage, ignored rows)
  - tolerance boundary behavior
  - missing-module-on-disk detection
  - `main()` hard-fail branches (missing report/baseline, malformed baseline JSON, unparseable report)
  - repo-root anchoring behavior independent of `cwd`
  - policy default resolution and invalid floor value handling

## Docs
- Updated coverage policy docs to correctly describe module scope as nested files under `src/file_organizer/`.

## Validation
- `pytest tests/ci/test_module_coverage_floor.py -q --override-ini='addopts='`
- `pre-commit run --files scripts/check_module_coverage_floor.py tests/ci/test_module_coverage_floor.py .github/workflows/ci.yml scripts/run-local-ci.sh scripts/coverage/README.md`

## Closes On Merge
Closes #31
Closes #32
Closes #33
Closes #34
Closes #35
Closes #37
Closes #38

## Follow-up
- #36 (`--update-baseline` write mode) is intentionally left as a separate enhancement to keep this PR focused on gate correctness, diagnostics, and parity.

## Summary by CodeRabbit

* **New Features**
  * Integration coverage now enforces per-module coverage floors alongside the global coverage threshold to prevent regressions in individual modules.

* **Documentation**
  * Added documentation outlining the two-layer coverage floor policy and process for ratcheting module coverage baselines.

* **Tests**
  * Added comprehensive test suite for module-level coverage validation logic.

* **Chores**
  * Updated CI workflows and local development scripts to implement enhanced per-module coverage enforcement.

### CODE REVIEW

File: scripts/check_module_coverage_floor.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        line = _strip_ansi(raw_line)
+        # GH Actions prefixes with tab-separated metadata; keep last segment.
+        line = line.split("\t")[-1].strip()
+        match = ROW_RE.search(line)
+        if not match:
+            continue
+        path = match.group(1)
+        cover = float(match.group(6))
+        modules[path] = cover
+    return modules
```
[the remark this answers is not in the document]
Comment (curdriceaurora) [reply scripts/check_module_coverage_floor.py:52]: Added in latest push (20acbbb): new focused tests in `tests/ci/test_module_coverage_floor.py` covering `parse_report()` and `evaluate()`.

Covered cases:
- parses GH Actions tab-prefixed lines + ANSI-colored rows
- ignores non-module rows
- flags regressions below baseline floor
- flags new modules below minimum
- flags baseline modules missing from report when file still exists
- verifies tolerance behavior for small drift

Validation run:
- `pytest tests/ci/test_module_coverage_floor.py -q --override-ini="addopts="`
- `pre-commit run --files tests/ci/test_module_coverage_floor.py scripts/check_module_coverage_floor.py`
